File: crates/yamind-commands/src/history.rs
use crate::command::Command;
use yamind_core::Document;

pub struct CommandHistory {
    undo_stack: Vec<Box<dyn Command>>,
    redo_stack: Vec<Box<dyn Command>>,
}

impl CommandHistory {
    pub fn new() -> Self {
        Self {
            undo_stack: Vec::new(),
            redo_stack: Vec::new(),
        }
    }

    pub fn execute(&mut self, mut cmd: Box<dyn Command>, doc: &mut Document) {
        cmd.execute(doc);
        self.undo_stack.push(cmd);
        self.redo_stack.clear();
    }

    /// Push a command that has already been executed (e.g. when we need to
    /// inspect the result before pushing to history).
    pub fn push_executed(&mut self, cmd: Box<dyn Command>) {
        self.undo_stack.push(cmd);
        self.redo_stack.clear();
    }

    pub fn undo(&mut self, doc: &mut Document) -> bool {
        if let Some(mut cmd) = self.undo_stack.pop() {
            cmd.undo(doc);
            self.redo_stack.push(cmd);
            true
        } else {
            false
        }
    }

    pub fn redo(&mut self, doc: &mut Document) -> bool {
        if let Some(mut cmd) = self.redo_stack.pop() {
            cmd.execute(doc);
            self.undo_stack.push(cmd);
            true
        } else {
            false
        }
    }

    /// Update the text on the last command in the undo stack.
    pub fn update_last_text(&mut self, text: String) {
        if let Some(cmd) = self.undo_stack.last_mut() {
            cmd.set_text(text);
        }
    }

    pub fn can_undo(&self) -> bool {
        !self.undo_stack.is_empty()
    }

    pub fn can_redo(&self) -> bool {
        !self.redo_stack.is_empty()
    }

    pub fn clear(&mut self) {
        self.undo_stack.clear();
        self.redo_stack.clear();
    }
}

impl Default for CommandHistory {
    fn default() -> Self {
        Self::new()
    }
}

File: src/src/shared/commands/history.ts
import type { Command } from './command'
import type { Document } from '../types/document'
import { isTextUpdatable } from './command'

export class CommandHistory {
  private undoStack: Command[] = []
  private redoStack: Command[] = []

  execute(cmd: Command, doc: Document): void {
    cmd.execute(doc)
    this.undoStack.push(cmd)
    this.redoStack = []
  }

  undo(doc: Document): boolean {
    const cmd = this.undoStack.pop()
    if (!cmd) return false
    cmd.undo(doc)
    this.redoStack.push(cmd)
    return true
  }

  redo(doc: Document): boolean {
    const cmd = this.redoStack.pop()
    if (!cmd) return false
    cmd.execute(doc)
    this.undoStack.push(cmd)
    return true
  }

  /** Update the text on the last command (for new-node commit workflow). */
  updateLastText(newText: string): boolean {
    const last = this.undoStack[this.undoStack.length - 1]
    if (!last || !isTextUpdatable(last)) return false
    last.updateText(newText)
    return true
  }

  get canUndo(): boolean {
    return this.undoStack.length > 0
  }

  get canRedo(): boolean {
    return this.redoStack.length > 0
  }

  get undoCount(): number {
    return this.undoStack.length
  }

  get redoCount(): number {
    return this.redoStack.length
  }
}
